PA-cma_microbenchmark.tex

```diff
@@ -0,0 +1,154 @@
+\subsection{CMA microbenchmark on Broadwell and TX2}
+\label{sec:pa-microbenchmark}
+In this subsection we will introduce the CMA\_appinv microbenchmark and present performance results from tests on the Intel Broadwell and Cavium ARM-based ThunderX2 architectures.
+In addition we compare performance of the microbenchmark when built with different compilers and runtime libraries.
+
+\subsubsection{CMA microbenchmark}
+
+The GungHo dynamical model performs operations to solve
+\begin{equation} \label{eq:matvec}
+A \cdot \mathbf{x} = \mathbf{b}
+\end{equation}
+in various guises and contexts where $A$ is a matrix representing a partial differential equation, $\mathbf{b}$ is a vector containing information describing boundary conditions and $\mathbf{x}$ is a state vector.
+The finite element methods that the GungHo application uses are written to perform operations on a cell-by-cell basis in what is termed Local Matrix Assembly (LMA).
+The LMA microbenchmark discussed in Section \ref{sec:pa} is a characteristic example of such methods.
+The LFRic infrastructure also contains methods for solving equation \ref{eq:matvec} on a column-by-column basis.
+Although these Column Matrix Assembly (CMA) methods have yet to be utilised by the full GungHo application they are deployed in the gravity\_wave mini-app.
+
+For this report we generated a new CMA microbenchmark using LFRic revision 15378 focused on the columnwise\_op\_appinv\_kernel\_code() routine in the gravity\_wave mini-app\footnote{At the time of writing the target kernel has not been significantly modified since extraction.}.
+This kernel is used in the pre-conditioner for the Helmholtz solver to solve the tridiagonal case of equation \ref{eq:matvec} using the Thomas Algorithm~\cite{numericalrecipes}.
+As per the LMA microbenchmark, the driver code was extracted from the Psyclone-generated Psy-layer and the kernel was taken directly from the LFRic source code.
+By design, the driver routine contains the so called global optimisations, namely OpenMP worksharing and colouring (to prevent data race conditions) over columns on the sphere.
+
+The dinodump capability from PYSKE was used to generate Known Good Output (KGO) which we use to verify the microbenchmark behaved as expected in later tests.
+This involved adding additional PSYKE subroutine calls to the Psy-layer code in the gravity\_wave mini-app in order to capture the input scalars and arrays for the kernel.
+The microbenchmark driver code was also trivially modified to read in the produced dinodump before calling the kernel and compare the result of the kernel against the KGO.
```

Comment:
Dinodumo should be dinodump

Reply:
Whoops. Done (in the web gui!) in commit 64726a5